### ISSUE TITLE
[Entitas:] Multithreading

### ISSUE THREAD

**Issue by sschmid** [sschmid/Entitas-CSharp#325]
# The future of Entitas - Multithreading
This is sth I wanted to work on for a long time and I think this can be a game changer!

### **The elevator pitch:**
Make Entitas ready for multithreading. Define best practises and workflows to enable everyone using Entitas to take full advantage of the given hardware and multiple processors.

---

Multithreading can be very hard and with Unity's limitation to work only on the main thread we lose a lot of potential! The ultimate goal is to simplify the workflow and turn multithreading into a no-brainer.

First results seem very promising

![entitas-threads](https://cloud.githubusercontent.com/assets/233700/23824485/fc368e1e-0677-11e7-9a40-08c398f04477.png)

**Comment by sschmid** [comment]
Feel free to retweet if you want to see this happening

https://twitter.com/entitas_csharp/status/840581685846634497

**Comment by WoLfulus** [comment]
what is the main problem with multithreading in entitas? concurrent access?

**Comment by sschmid** [comment]
All common pitfalls when using MT also apply when using Entitas. My main goal is to explore and define workflows and conventions to help developers avoid having issues. Those conventions could lead to additional apis to encourage people to structure their code in a certain way. After that, the ultimate goal which I'd love to achieve somehow is that you don't even have to care about details of MT; Entitas could figure out those details for you and run systems in parallel by itself. But that would be not part of Phase, but I like to set the goals high ;)

**Comment by paraboxx** [comment]
There is an older ECS framework for .NET on github called Forge (https://github.com/jacobdufault/forge) which implemented MT by enforcing immutable state. The whole architecture basically uses double buffering for the gamestate. During one update cycle, the systems read only from the old state and only write to the next. No interdependencies between state changes in a single update cycle are allowed. I didn't use it much though and sadly it didn't pay any attention to cache coherency and data locality.

**Comment by danielbe1** [comment]
If I take Entitas-Shmup as a case for MT, most of the systems there could run code in parallel on the Entity level, for example updating positions, velocity, damage and all that. Basically, any processing that does not depend on another entity's components can run in this mode.

Running systems in parallel is not as simple, but because entities are a composition of many components. systems that operate on separate components could run in parallel. It would be nice if we had systems that operated on subsets of the components and combined the results at the end somehow. Maybe under the existing code, primitives like EntitySystem and ComponentSystems could be used to run code in parallel.

**Comment by periodyctom** [comment]
An interesting approach might be something similar to what this guy is doing with a C++ ECS framework:
- [Vittorio Romeo: Implementation of a multithreaded compile-time ECS in C++14](https://www.youtube.com/watch?v=3N1pLtTV2Uc)
- [Vittorio Romeo's ECST](https://github.com/SuperV1234/ecst)

While he uses C++ templates to setup the metadata and configure systems, I like the idea of having predefined "modes" for parallelization. There would also be the option to customize for specialized use cases, as needed. His technique also ensures that systems later in the chain must wait for all of their dependencies to complete before executing, which seems to remove the need for double buffering.

**Comment by KumoKairo** [comment]
https://youtu.be/YHweZ8dhOJA?t=4483
It may go side by side with the new Unity C# job system (safe multithreading)

**Comment by periodyctom** [comment]
@KumoKairo - That's what I was thinking, assuming the Jobs system comes out at a reasonable pace.

**Comment by aefreedman** [comment]
Just throwing in here that there might be a conflict with WebGL builds and multithreading.

**Comment by VladislavKostin** [comment]
I actually don't see how Job system would help Entitas, I assume it will have some overhead. It seems to me that some solution like double buffering is the easiest one to work with and it would "just work".

In Entitas I imagine you could mark systems as multithreaded, and they would be completely multithreaded under the hood. Entitas would run multithreaded systems, then in the main thread swap the buffers, run through systems that can't be multithreaded for some game logic reasons, start multithreaded systems again, and run all view systems in the main thread meanwhile.

@periodyctom Parallelisation without the double buffering? Is that doable for Entitas? Would determinism be sacrificed? Would reduced memory footprint be the only gain compared to double buffer?

**Comment by periodyctom** [comment]
@VladislavKostin Honestly, you're probably right about double buffering being the simplest, I'm merely pointing out that I like the policy based approach presented in the links I posted above, for the people who need more control for specialized scenarios. The idea of customizing the threading work chain (with some rules to enforce determinism) is very appealing.

What I would like to see is the ability to setup the threading policy for systems, with double-buffering probably being the "default, simple, most tested" option, but having the ability to say "run the system chain synchronously, but for Feature/System X, split the entity workload across all cores, then resume on the main thread", perhaps with some stipulations or deferred operations like in the example I posted.

**Comment by Lynx32** [comment]
Hello @sschmid,
we are experimenting with Entitas right now. I was suprised to discover that using Entitias would make code significantly harder to multithread. That is our main concern with Entitas now. We are currently considering modifying Entitias to add some form of double buffering or sadly discarding it.
I was really happy to read about your plans here. I am curious how would you like to achieve it. I guess you have some thoughts already about that. Could your shed some light on it and share the rough ideas of that practices and workflows you have came up with? Also, is there any time horizon at which you expect it to be ready?

**Comment by T2RKUS** [comment]
surely we can use code similar to this and subclass ReactiveSystem -> ParallelReactiveSystem? 
https://github.com/thelinuxlich/artemis_CSharp/blob/master/Artemis_XNA_INDEPENDENT/System/ParallelEntityProcessingSystem.cs

Or am I missing something?

**Comment by T2RKUS** [comment]
I wanted to get everyone's thoughts on this before implementing it myself and trying it out, don't want to waste my efforts if it's not going to work (perhaps the component/context events triggering in parallel cause unintended side effects. not sure how that works.)

**Comment by optimisez** [comment]
I will suggest waiting for Unity C# job system coming.

**Comment by VladislavKostin** [comment]
I'm wondering about the future of Entitas now that Unity releasing it's own ECS.

Unity's new ECS + Job System + Compiler Tech is an insanely optimized system.  They claim that their code is even faster than vectorized C++ code, because of their custom compiler optimizations. And this will be available only in Unity. They have manual memory management with all it's drawbacks and performance advantages with zero garbage allocation. That's gonna be the "right" way to create games, and all advanced Unity developers will learn it. Unity has almost 100% multicore utilization.

Entitas probably won't be able to reach the same level of performance. Entitas might have better API though and better workflow now with compiler. But then Unity ECS probably gonna have better support, docs and tutorials. Most of the developers don't look for super high performance, they just want a good architecture and API that are easy to work with. Also Entitas is portable and can be stand alone. The reactive systems might be an optimization (is it?) that Unity doesn't have (at least yet).

I'm not even sure what I would choose, of course, depends on a project as well.

**Comment by T2RKUS** [comment]
True but not everyone will be using unity so for instances when you don't want to use unity but want to an ECS framework you'd be glad Entitas exists whether you're making a game or not. :)

**Comment by KumoKairo** [comment]
@T2RKUS I would go and implement this thing anyway. After these experiments we will at least know current friction points of multithreading with Entitas and will know where to move next.
As for the Unity's ECS - their solution is likely to be very specific, while general multithreading is usually more general and applicable to more usage cases, so it will find its user even after Unity will release their super-fast-and-specific ECS

**Comment by Lynx32** [comment]
@T2RKUS
Generally your idea of "ParallelReactiveSystem" is valid and it would be a good and easy way to multithread stuff to some degree. If it fits your needs then go ahead. When writing my post I thought about higher level parallelization tools, something like upcoming Unity Job system with job dependencies and other stuff.

**Comment by KumoKairo** [comment]
@Lynx32 it's been a while since June 1st, can you share your experience with Entitas so far? Did you decide to discard it or still using it?

**Comment by Lynx32** [comment]
@KumoKairo We have used entitas and we are glad we did it. I really enjoy working with it. Unfortunately, as the game is on a prototyping stage, we still did not bother to adapt entitas for our needs in MT. We have reviewed important parts of the source code and prepared "strategy" for implementing it. We also keep annotating dependencies between the systems and try to keep code ready for the change, but that is all for now in this matter.

It is hard to tell when I will get to make that change. It will not be sooner than January 2018. 
I also really look forward to Unity Job System and Unity ECS. Depending on the time of its introduction and state of development of our game, we would also consider switching to Unity solution.

**Comment by sschmid** [comment]
I played around with threading again. I added a [JobSystem.cs](https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Systems/JobSystem.cs) as a proof of concept.

This system distributes the execute over multiple threads and joins them again. This can speed up performance if you either have lots of entities or if the calculations are very intense.

It's really simple to set up:

```csharp
public sealed class RotateSystem : JobSystem<GameEntity> {

    public RotateSystem(GameContext context, int threads) :
        base(context.GetGroup(GameMatcher.AllOf(GameMatcher.Rotation, GameMatcher.RotationVector)), threads) {
    }

    protected override void Execute(GameEntity entity) {
        entity.rotation.value = entity.rotation.value * Quaternion.Euler(entity.rotationVector.value);
    }
}
```

![entitas-perormance-comparisson](https://user-images.githubusercontent.com/233700/37561147-9414d742-2a47-11e8-9399-51813c0287a1.png)

**Comment by SylwesterMichalowski** [comment]
hi, I have one question the JobSystem.cs is still proof of concept?

**Comment by rikkuporta** [comment]
HI, how is this gping along? Is this on par with Unitys ECS/Jobs?
Cant really decide which to use for my next project

I guess Burst is out of question tho.

**Comment by sschmid** [comment]
@rikkuporta 

If you have high requirements and you already know that you have to multithread big parts of your simulation, try Unity ECS.
Entitas currently has a JobSystems class which split the task on multiple cores, but has limitations similar to Unity ECS where you have to delay entity creation and deletion.
